lib/src/erc20/erc20_base_subsystem.cairo

```diff
@@ -0,0 +1,131 @@
+#[contract]
+mod ERC20BaseSubsystem {
+
+    use dojo::world;
+    use zeroable::Zeroable;
+    use starknet::get_caller_address;
+    use starknet::contract_address_const;
+    use starknet::ContractAddress;
+    use starknet::ContractAddressZeroable;
+
+    struct Storage {
+        name: felt252,
+        symbol: felt252,
+        decimals: u8,
+        total_supply: u256,
+        balances: OwnershipComponent,
+        allowances: ApprovalComponent,
+    }
+
+    #[event]
+    fn Transfer(from: ContractAddress, to: ContractAddress, value: u256) {}
+
+    #[event]
+    fn Approval(owner: ContractAddress, spender: ContractAddress, value: u256) {}
+
+    #[constructor]
+    fn constructor(
+        name_: felt252,
+        symbol_: felt252,
+        decimals_: u8,
+        initial_supply: u256,
+        recipient: ContractAddress
+    ) {
+        name::write(name_);
+        symbol::write(symbol_);
+        decimals::write(decimals_);
+        assert(!recipient.is_zero(), 'ERC20: mint to the 0 address');
+        total_supply::write(initial_supply);
+        balances::write(recipient, initial_supply);
+        Transfer(contract_address_const::<0>(), recipient, initial_supply);
+    }
+
+    #[view]
+    fn get_name() -> felt252 {
+        name::read()
+    }
+
+    #[view]
+    fn get_symbol() -> felt252 {
+        symbol::read()
+    }
+
+    #[view]
+    fn get_decimals() -> u8 {
+        decimals::read()
+    }
+
+    #[view]
+    fn get_total_supply() -> u256 {
+        total_supply::read()
+    }
+
+    #[view]
+    fn balance_of(account: ContractAddress) -> u256 {
+        balances::read(account)
+    }
+
+    #[view]
+    fn allowance(owner: ContractAddress, spender: ContractAddress) -> u256 {
+        allowances::read((owner, spender))
+    }
+
+    #[external]
+    fn transfer(recipient: ContractAddress, amount: u256) {
+        let sender = get_caller_address();
+        transfer_helper(sender, recipient, amount);
+    }
+
+    #[external]
+    fn transfer_from(sender: ContractAddress, recipient: ContractAddress, amount: u256) {
+        let caller = get_caller_address();
+        spend_allowance(sender, caller, amount);
+        transfer_helper(sender, recipient, amount);
+    }
+
+    //approval system
+    #[external]
+    fn approve(spender: ContractAddress, amount: u256) {
```

Comment:
The logic for  `transfer_from` and `approve` need to be moved to systems.

For example approval could look something like this:

```
#[system]
mod ERC20_Approve {
    execute(token_id: felt252, spender: ContractAddress, amount: u256) {
        let caller = get_caller_address();
        let approval_sk: StorageKey = (token_id, (caller, spender)).into();
        let approval = commands::<Approval>::get(approval_sk);
        commands::set(approval_sk, (
            Approval { amount: approval.amount + amount }
        ))
    }
}
```